src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1253 @@
+                /// This makes it easier to track input types while processing the eval sequence.
+                /// </summary>
+                public bool? OnTheLeftOfDisjunction { get; set { Debug.Assert(field is null); field = value; } }
+
+                private OperandOrOperation(BoundPattern? operand, bool? disjunction, SyntaxNode? operationSyntax)
+                {
+                    _operand = operand;
+                    OnTheLeftOfDisjunction = null;
+                    _disjunction = disjunction;
+                    _operationSyntax = operationSyntax;
+                }
+
+                public static OperandOrOperation CreateOperation(bool disjunction, SyntaxNode operationSyntax)
+                {
+                    return new OperandOrOperation(null, disjunction, operationSyntax);
+                }
+
+                public static OperandOrOperation CreateOperand(BoundPattern operand)
+                {
+                    return new OperandOrOperation(operand, null, null);
+                }
+
+                public bool IsOperand([NotNullWhen(true)] out BoundPattern? operand)
+                {
+                    if (_operand is not null)
+                    {
+                        operand = _operand;
+                        return true;
+                    }
+
+                    operand = null;
+                    return false;
+                }
+
+                public bool IsOperation(out bool disjunction, [NotNullWhen(true)] out SyntaxNode? operationSyntax)
+                {
+                    if (_operand is null)
+                    {
+                        Debug.Assert(_disjunction.HasValue);
+                        disjunction = _disjunction.Value;
+                        Debug.Assert(_operationSyntax is not null);
+                        operationSyntax = _operationSyntax;
+                        return true;
+                    }
+
+                    disjunction = false;
+                    operationSyntax = null;
+                    return false;
+                }
+            }
+
+            private PatternNormalizer()
+            {
+            }
+
+            internal static BoundPattern Rewrite(BoundPattern pattern, TypeSymbol inputType)
+            {
+                var patternNormalizer = new PatternNormalizer();
+                patternNormalizer.Visit(pattern);
+                return patternNormalizer.GetResult(inputType);
+            }
+
+            private BoundPattern GetResult(TypeSymbol inputType)
+            {
+                Debug.Assert(_evalSequence is [var first, ..] && first.IsOperand(out _));
+
+                var stack = ArrayBuilder<BoundPattern>.GetInstance();
+
+                int evalPosition = 0;
+                do
+                {
+                    switch (_evalSequence[evalPosition])
+                    {
+                        case var x when x.IsOperation(out bool disjunction, out SyntaxNode? operationSyntax):
```

Comment:
>var x when x.IsOperation(out bool disjunction, out SyntaxNode? operationSyntax): [](http://example.com/codeflow?start=29&length=80)

After this change, there is no any advantage in using `switch`, all meaningful tests are in `when` clauses. Therefore, I suggest switching to regular `if` statements and getting rid of the "dummy" ```var x/y``` patterns.